src/Collapse/include/gudhi/Flag_complex_sparse_matrix.h

```diff
@@ -0,0 +1,424 @@
+ */
+
+#ifndef FLAG_COMPLEX_SPARSE_MATRIX_H_
+#define FLAG_COMPLEX_SPARSE_MATRIX_H_
+
+#include <gudhi/graph_simplicial_complex.h>
+
+#include <boost/functional/hash.hpp>
+#include <boost/graph/adjacency_list.hpp>
+
+#include <Eigen/Sparse>
+
+#ifdef GUDHI_USE_TBB
+#include <tbb/parallel_sort.h>
+#endif
+
+#include <iostream>
+#include <utility>  // for std::pair
+#include <vector>
+#include <unordered_map>
+#include <unordered_set>
+#include <set>
+#include <tuple>  // for std::tie
+#include <algorithm>  // for std::includes
+#include <iterator>  // for std::inserter
+
+namespace Gudhi {
+
+namespace collapse {
+
+/**
+ * \class Flag_complex_sparse_matrix
+ * \brief Flag complex sparse matrix data structure.
+ *
+ * \ingroup collapse
+ *
+ * \details
+ * This class stores a <a target="_blank" href="https://en.wikipedia.org/wiki/Clique_complex">Flag complex</a>
+ * in an <a target="_blank" href="https://eigen.tuxfamily.org/dox/group__TutorialSparse.html">Eigen sparse matrix</a>.
+ *
+ * \tparam Vertex type must be a signed integer type. It admits a total order <.
+ * \tparam Filtration type for the value of the filtration function. Must be comparable with <.
+ */
+template<typename Vertex, typename Filtration>
+class Flag_complex_sparse_matrix {
+ public:
+  /** \brief Re-define Vertex as Vertex_handle type to ease the interface with compute_proximity_graph. */
+  using Vertex_handle = Vertex;
+  /** \brief Re-define Filtration as Filtration_value type to ease the interface with compute_proximity_graph. */
+  using Filtration_value = Filtration;
+  /** \brief This is an ordered pair, An edge is stored with convention of the first element being the smaller i.e
+   * {2,3} not {3,2}. However this is at the level of row indices on actual vertex lables.
+   */
+  using Edge = std::pair<Vertex_handle, Vertex_handle>;
+
+ private:
+  // Row_index type in the sparse matrix
+  using Row_index = std::size_t;
+
+  // The sparse matrix data type
+  using Sparse_row_matrix = Eigen::SparseMatrix<Filtration_value, Eigen::RowMajor>;
+
+  // A range of row indices
+  using Row_indices_vector = std::vector<Row_index>;
+
+ public:
+  /** \brief Filtered_edge is a type to store an edge with its filtration value. */
+  using Filtered_edge = std::pair<Edge, Filtration_value>;
+  /** \brief Proximity_graph is a type that can be used to construct easily a Flag_complex_sparse_matrix. */
+  using Proximity_graph = Gudhi::Proximity_graph<Flag_complex_sparse_matrix>;
+
+ private:
+  // Map from row index to its vertex handle
+  std::unordered_map<Row_index, Vertex_handle> row_to_vertex_;
+
+  // Vertices stored as an unordered_set
+  std::unordered_set<Vertex_handle> vertices_;
+
+  // Unordered set of removed edges. (to enforce removal from the matrix)
+  std::unordered_set<Edge, boost::hash<Edge>> u_set_removed_edges_;
+
+  // Unordered set of dominated edges. (to inforce removal from the matrix)
+  std::unordered_set<Edge, boost::hash<Edge>> u_set_dominated_edges_;
```

Comment:
I am trying to understand if this is redundant with critical_edge_indicator_ or not. It seems that filling this structure is what prevents us in set_edge_critical from iterating on effected_indices, instead of going through all edges.

Reply:
I removed it on 85dc5c1e
It is "strange" as I have to catch exception to know if it is present or not.
Do not hesitate if you want me to rollback.

Reply:
```c++
      if (rw_u != rw_v) {
        bool b1 = u_set_dominated_edges_.count(std::minmax(rw_u, rw_v)) != 0;
        Row_index ei = edge_to_index_map_.at(std::minmax(row_to_vertex_[rw_u], row_to_vertex_[rw_v]));
        bool b2 = ei > current_backwards && !critical_edge_indicator_[ei];
        assert(b1==b2);
      }
```
does not fail while running the tests.